ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/ks/KsMaterializedSessionTableIQv2.java

```diff
@@ -0,0 +1,138 @@
+import io.confluent.ksql.execution.streams.materialization.MaterializedWindowedTable;
+import io.confluent.ksql.execution.streams.materialization.WindowedRow;
+import java.time.Instant;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.Window;
+import org.apache.kafka.streams.kstream.Windowed;
+import org.apache.kafka.streams.query.QueryResult;
+import org.apache.kafka.streams.query.StateQueryRequest;
+import org.apache.kafka.streams.query.StateQueryResult;
+import org.apache.kafka.streams.query.WindowRangeQuery;
+import org.apache.kafka.streams.state.KeyValueIterator;
+
+/**
+ * Kafka Streams impl of {@link MaterializedWindowedTable}.
+ */
+class KsMaterializedSessionTableIQv2 implements MaterializedWindowedTable {
+
+  private final KsStateStore stateStore;
+
+  KsMaterializedSessionTableIQv2(final KsStateStore store) {
+    this.stateStore = Objects.requireNonNull(store, "store");
+  }
+
+  @Override
+  public List<WindowedRow> get(
+      final GenericKey key,
+      final int partition,
+      final Range<Instant> windowStart,
+      final Range<Instant> windowEnd
+  ) {
+    try {
+      return findSession(key, partition, windowStart, windowEnd);
+    } catch (final Exception e) {
+      throw new MaterializationException("Failed to get value from materialized table", e);
+    }
+  }
+
+  @Override
+  public Iterator<WindowedRow> get(final int partition, final Range<Instant> windowStartBounds,
+      final Range<Instant> windowEndBounds) {
+    throw new MaterializationException("Table scan unsupported on session tables");
+  }
+
+  private List<WindowedRow> findSession(
+      final GenericKey key,
+      final int partition,
+      final Range<Instant> windowStart,
+      final Range<Instant> windowEnd
+  ) {
+
+    final WindowRangeQuery<GenericKey, GenericRow> query = WindowRangeQuery.withKey(key);
+    final StateQueryRequest<KeyValueIterator<Windowed<GenericKey>, GenericRow>> request =
+        inStore(stateStore.getStateStoreName()).withQuery(query);
+    final StateQueryResult<KeyValueIterator<Windowed<GenericKey>, GenericRow>> result =
+        stateStore.getKafkaStreams().query(request);
+
+    final Map<Integer, QueryResult<KeyValueIterator<Windowed<GenericKey>, GenericRow>>>
+        partitionResults = result.getPartitionResults();
+
+    if (partitionResults.size() != 1) {
+      throw new MaterializationException("The query did not return exactly one partition result: "
+        + result.getPartitionResults());
+    }
+
+    if (!partitionResults.containsKey(partition)) {
+      throw new MaterializationException("The query did not return the requested partition result: "
+        + result.getPartitionResults());
+    }
+
+    final QueryResult<KeyValueIterator<Windowed<GenericKey>, GenericRow>> queryResult =
+        partitionResults.get(partition);
```

Comment:
```suggestion
    final QueryResult<KeyValueIterator<Windowed<GenericKey>, GenericRow>> queryResult =
        result.getPartitionResults().get(partition);
```

See https://github.com/confluentinc/ksql/pull/8570